crypto/bls_thresholdsign.go

```diff
@@ -0,0 +1,594 @@
+// If no error is returned, the bool represents the validity of the signature.
+// The function is thread-safe and non-blocking.
+func (s *blsThresholdSignatureFollower) VerifyShare(orig int, share Signature) (bool, error) {
+	// validate index
+	if err := s.validIndex(index(orig)); err != nil {
+		return false, err
+	}
+
+	return s.publicKeyShares[orig].Verify(share, s.message, s.hasher)
+}
+
+// VerifyThresholdSignature verifies the input signature against the stored
+// message and stored group public key.
+//
+// This function does not update the internal state.
+// The function errors if the execution failed.
+// The function does not return an error for any invalid signature.
+// If any error is returned, the returned bool is false.
+// If no error is returned, the bool represents the validity of the signature.
+// The function is thread-safe and non-blocking only.
+func (s *blsThresholdSignatureFollower) VerifyThresholdSignature(thresholdSignature Signature) (bool, error) {
+	return s.groupPublicKey.Verify(thresholdSignature, s.message, s.hasher)
+}
+
+// EnoughShares checks whether there are enough shares to reconstruct a signature.
+// The funstion returns true if and only if the number of shares have reached (threshold+1)
+// shares.
+//
+// This function is thread safe and write-blocking
+func (s *blsThresholdSignatureFollower) EnoughShares() bool {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	return s.enoughShares()
+}
+
+// non thread safe version of EnoughShares
+func (s *blsThresholdSignatureFollower) enoughShares() bool {
+	// len(s.signers) is always <= s.threshold + 1
+	return len(s.shares) == (s.threshold + 1)
+}
+
+// HasShare checks whether the internal map contains the share of the given index.
+// This function is thread safe and write-blocking
+func (s *blsThresholdSignatureFollower) HasShare(orig int) (bool, error) {
+	// validate index
+	if err := s.validIndex(index(orig)); err != nil {
+		return false, err
+	}
+
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	return s.hasShare(index(orig)), nil
+}
+
+// non thread safe version of HasShare, and assumes input is valid
+func (s *blsThresholdSignatureFollower) hasShare(orig index) bool {
+	_, ok := s.shares[orig]
+	return ok
+}
+
+// duplicatedSignerError is an error returned when TrustedAdd or VerifyAndAdd encounter
+// a signature share that has been already added to the internal state.
+type duplicatedSignerError struct {
+	message string
+}
+
+// duplicatedSignerErrorf constructs a new duplicatedSignerError
+func duplicatedSignerErrorf(msg string, args ...interface{}) error {
+	return &duplicatedSignerError{message: fmt.Sprintf(msg, args...)}
+}
+
+func (e duplicatedSignerError) Error() string {
+	return e.message
+}
+
+// IsduplicatedSignerError checks if the input error is a duplicatedSignerError
+func IsduplicatedSignerError(err error) bool {
+	var target *duplicatedSignerError
+	return errors.As(err, &target)
+}
```

Comment:
Since the package covers many algorithms/protocols, and this error is related only to threshold signatures, I prefer that it remains close the threshold signature code. I have moved it near the interface file, instead of the implementation one.